### PR TITLE
Feature/set state and device handler

### PR DESCRIPTION
**Pull request by gandazgul**
This is adding the ability for sending set_state messages from MQTT to SmartThings and have it setStatus on a device that supports it. 

I'm also adding here a Garage Door Opener based on the simulated one from SmartThings that supports setting status. I'm not 100% sure on this one, I couldn't find a way to set the status of the door without sending an event. So maybe is not worth doing setStatus but rather calling open() and still ignoring the resulting event.

### CODE REVIEW

**Comment by john3exonets** [comment]
+1 for this kind of feature.  I have lots of sensors at my house that use MQTT to report.....it would be nice to have ST be able to see the data coming over the bridge and update psuedo-devices.  I know its not all do-able with this bridge code, but this at least gets the data inside ST.

**Comment by gandazgul** [comment]
@john3exonets What else are you trying to do that you don't think is doable? I'm getting great results with this bridge.

File: smartapps/stj/mqtt-bridge.src/mqtt-bridge.groovy
Comment (stjohnjohnson) [inline smartapps/stj/mqtt-bridge.src/mqtt-bridge.groovy]: Is there a race condition between this line and the line above?

**Review by stjohnjohnson — CHANGES_REQUESTED** [review]
Can you please update the README as well?

**Comment by stjohnjohnson** [comment]
Btw, this is really excellent.  I like the changes here.  Not sure what to do about that garage door opener.

File: smartapps/stj/mqtt-bridge.src/mqtt-bridge.groovy
Comment (gandazgul) [reply smartapps/stj/mqtt-bridge.src/mqtt-bridge.groovy]: It shouldn't as they should happen in 2 different cycles. All of this event handler should happen in one cycle, then any changes to the device being reported should happen on the next, I could be wrong about this but I have been using both the garage opener and a light using this and haven't observed a race condition.

**Comment by gandazgul** [comment]
@stjohnjohnson thanks for reviewing these changes. I will update the readme now.

As for the garage opener device, I'm also working on a device to control a ceiling fan with this bridge. I figured we can have them here so that people get a complete package. I'm open to writing/help write other device handlers as needed. If you really think it doesn't belong I'll create a new repo under my profile and store it there with links to his repo.

**Comment by gandazgul** [comment]
@stjohnjohnson I updated the readme, let me know how you like it.

**Comment by airdrummingfool** [comment]
First off, thank you for writing this feature! I've been wishing for exactly this (it looks to be a better/more generic solution than I came up with in #82).

I've been watching this PR for a while, and I'd like to add my (humble and unsolicited) two cents regarding the garage door opener DTH: I don't think miscellaneous device handlers belong here - I think setting up your own repo and linking to this in the install instructions would be a better way to go.

If this repo becomes a home for every device handler that uses MQTT, the issue and PR queues will become littered with people who need support for X handler, instead of the MQTT bridge code itself. Even worse, whoever actually maintains handlers committed to this repository wouldn't even be able to be easily/reliably notified when an issue pertaining to their DTH was raised. Committing bug fixes or new features for a handler would require @stjohnjohnson's explicit approval, or else every person maintaining a DTH would need commit access to this repo.

All that being said, I do think it would be nice to have some kind of directory or collection of virtual (MQTT-compatible) device handlers. I have written a couple myself, but they haven't been put in a repository yet.

Thanks again for this very useful addition!

**Comment by gandazgul** [comment]
@airdrummingfool thanks! I didn't consider the maintenance implications, you are absolutely right. I will then move the garage door device and my other ceiling fan I'm making to a separate repo. I'll update the PR tonight when I get home.

**Comment by stjohnjohnson** [comment]
I agree with @airdrummingfool's comments.  I'm happy to provide links to other MQTT-based DTHs.

Also, one last comment, can you update the default config template: https://github.com/stjohnjohnson/smartthings-mqtt-bridge/blob/master/_config.yml

**Comment by gandazgul** [comment]
@stjohnjohnson Updated config template

**Review by stjohnjohnson — APPROVED** [review]
Awesome work!  Thanks again!

**Comment by stjohnjohnson** [comment]
Because you removed `state_suffix`, I bumped the version to v2.0.0 and published it in NPM and soon Docker Hub.  Thanks again for the PR.

**Comment by gandazgul** [comment]
Woohoo! Awesome. Thanks for merging this!

**Comment by hshir** [comment]
Hi Guys. 
I'm sorry but I've got lost here......I'll appreciate some help.
I'm not a programmer and don't have much experience here ,regrading understanding how to track the files updates, I just took the latest from the github page.
Anyway, I have the bridge setup and working OK on my ST device.
The things is' I can't get status messages back to ST.
For example, I have a MQTT temperature sensor I want to update on a virtual sensor on ST.
What do I need to update on ST?
Currently I have the latest DH and Smartapp updated, but probably something is missing to get it working.
Appreciate your help guys.

**Comment by gandazgul** [comment]
For this particular change to work the device has to have implemented a setStatus command that will be called with the key and value to set. Smartthings doesn't natively provide a way to set states without commands. Normally sensors don't have commands to set their state because they are the source of the measurement if that makes sense.

What you need to do is make your own device handler. Take the virtual sensor code as a starting point and add a setStatus command. See an example here: https://github.com/gandazgul/mqtt-smartthings-devices/blob/master/devicetypes/gandazgul/mqtt-ceiling-fan.src/mqtt-ceiling-fan.groovy

**Comment by gandazgul** [comment]
@hshir All you need is a temp sensor? I'll make the device handler and post it to that repo above. That way you can install it using the GitHub integration.

**Comment by hshir** [comment]
@gandazgul 
I really appreciate your help, thx alot for the detailed explanation , yes it does make sense, even though
I don't understand why for example virtual presence sensor does work and temperature sensor doesn't,
Anyways, for now I just need the temperature sensor, I appreciate if your can post it to your repo.

Just one more question regarding that, do I need to update something in the Bridge device handler and/or the Bridge smartapp in addition to the virtual temp sensor DH?
If yes, can you send a link for the repo for those as well, since I have seen so many updates and corrections people made,and I've really got lost.

Thx alot man for your help :-)

**Comment by gandazgul** [comment]
@hshir it really depends on whether the sensor device handler implements a set command for that capability. It possible for the presence sensor they did write in the setPresence command. 

You wouldn't need to make changes anywhere else only use my device handler. I'll try and post it tomorrow.

**Comment by hshir** [comment]
@gandazgul.

Appreciate your help 😊, thx.
Please, Let me know when it's posted 🙏

**Comment by gandazgul** [comment]
@hshir Here you go, sorry I didn't post it sooner I've been swamped this week. 

https://github.com/gandazgul/mqtt-smartthings-devices/blob/master/devicetypes/gandazgul/mqtt-temperature-sensor.src/mqtt-temperature-sensor.groovy

I tested it and I'm able to set the temp and humidity with these:

Write to 
`smartthings/DEVICE NAME/temperature/set_state "70"`
and
`smartthings/DEVICE NAME/humidity/set_state "35"`

I added a health tile as well, I'm working on a PR for this bridge to add Heath as a set/get capability, I'm not sure if its possible yet.

**Pull request merged**
